server/src/main/java/org/apache/druid/server/coordinator/duty/KillUnusedSegments.java

```diff
@@ -146,130 +156,128 @@ public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
+        remainingDatasourcesToKill.remove(dataSource);
+      }
+      catch (Exception ex) {
+        log.error(ex, "Failed to submit kill task for dataSource[%s] in interval[%s]", dataSource, intervalToKill);
+        if (Thread.currentThread().isInterrupted()) {
+          log.warn("Skipping kill task scheduling because thread is interrupted.");
+          break;
         }
       }
     }
 
-    if (log.isDebugEnabled()) {
-      log.debug(
-          "Submitted [%d] kill tasks for [%d] datasources.%s",
-          submittedTasks,
-          dataSourcesToKill.size(),
-          availableKillTaskSlots < dataSourcesToKill.size()
-              ? StringUtils.format(
-              " Datasources skipped: %s",
-              ImmutableList.copyOf(dataSourcesToKill).subList(submittedTasks, dataSourcesToKill.size())
-          )
-              : ""
-      );
-    }
+    log.info(
+        "Submitted [%d] kill tasks for [%d] datasources. Remaining datasources to kill: %s",
+        submittedTasks,
+        dataSourcesToKill.size(),
+        remainingDatasourcesToKill
+    );
 
-    // report stats
-    return submittedTasks;
+    stats.add(Stats.Kill.SUBMITTED_TASKS, submittedTasks);
   }
 
   /**
+   * <p>
    * Calculates the interval for which segments are to be killed in a datasource.
+   * Because this function uses string comparisons for date time, it doesn't find unused segments that are outside
+   * the range [{@link DateTimes#COMPARE_DATE_AS_STRING_MIN}, {@link DateTimes#COMPARE_DATE_AS_STRING_MAX}),
+   * such as {@link org.apache.druid.java.util.common.granularity.Granularities#ALL} partitioned segments and segments that
+   * end in {@link DateTimes#MAX}.
+   *</p><p>
+   * For more information, see <a href="https://github.com/apache/druid/issues/15951"> Issue#15951</a>.
+   * </p>
    */
   @Nullable
-  private Interval findIntervalForKill(String dataSource, DateTime maxUsedStatusLastUpdatedTime)
+  private Interval findIntervalForKill(
+      final String dataSource,
+      final DateTime maxUsedStatusLastUpdatedTime,
+      final CoordinatorRunStats stats
+  )
   {
     final DateTime maxEndTime = ignoreDurationToRetain
                                 ? DateTimes.COMPARE_DATE_AS_STRING_MAX
                                 : DateTimes.nowUtc().minus(durationToRetain);
-    List<Interval> unusedSegmentIntervals = segmentsMetadataManager
-        .getUnusedSegmentIntervals(dataSource, datasourceToLastKillIntervalEnd.get(dataSource), maxEndTime, maxSegmentsToKill, maxUsedStatusLastUpdatedTime);
+
+    final List<Interval> unusedSegmentIntervals = segmentsMetadataManager.getUnusedSegmentIntervals(
+        dataSource,
+        datasourceToLastKillIntervalEnd.get(dataSource),
+        maxEndTime,
+        maxSegmentsToKill,
+        maxUsedStatusLastUpdatedTime
+    );
+
+    // Each unused segment interval returned here has a 1:1 correspondence with an unused segment. So we can assume
```

Comment:
How do we know this? A single interval can have many killable unused segments.

Reply:
Please see `getUnusedSegmentIntervals()` - it returns a list of unused segment intervals when it could actually just be the unused segments itself, but more lightweight computation. Also, added `testKillMultipleSegmentsInSameInterval()` to demonstrate the scenario.

I considered changing it to directly fetch the unused segments instead of the intervals, which would also allow us to compute any additional statistics like size, etc. If you think that's useful, I can do it in a follow-up patch.

Reply:
I see, thanks for the clarification! Looking at the implementation `SqlSegmentsMetadataManager`, that does indeed seem to be the case. I guess the javadoc in `SegmentsMetadataManager` didn't really call it out and I just assumed that it would be a list of unique intervals.

Although, it is a little weird for this API to return duplicate intervals, maybe something we should revisit? Or atleast call out in the `SegmentsMetadataManager` javadocs for now?

Reply:
Yeah, I think we can revisit that later. Let me clarify the current behavior in the method's javadocs to avoid any confusion